pkg/mesh/mesh.go

```diff
@@ -774,19 +795,31 @@ func linkByIndex(index int) (netlink.Link, error) {
 
 // updateNATEndpoints ensures that nodes and peers behind NAT update
 // their endpoints from the WireGuard configuration so they can roam.
-func updateNATEndpoints(nodes map[string]*Node, peers map[string]*Peer, conf *wireguard.Conf) {
+func updateNATEndpoints(nodes map[string]*Node, peers map[string]*Peer, conf *wireguard.Conf, logger log.Logger) map[string]*wireguard.Endpoint {
+	natEndpoints := make(map[string]*wireguard.Endpoint)
 	keys := make(map[string]*wireguard.Peer)
 	for i := range conf.Peers {
 		keys[string(conf.Peers[i].PublicKey)] = conf.Peers[i]
 	}
 	for _, n := range nodes {
 		if peer, ok := keys[string(n.Key)]; ok && n.PersistentKeepalive > 0 {
-			n.Endpoint = peer.Endpoint
+			level.Debug(logger).Log("msg", "WireGuard Update NAT Endpoint", "node", n.Name, "endpoint", peer.Endpoint, "former-endpoint", n.Endpoint, "same", n.Endpoint.Equal(peer.Endpoint))
+			// Only public ip ? (should check location leader but only in topology ... or have topology handle that list)
+			// Better check wg latest-handshake
```

Comment:
So the idea is to only set the endpoint, if the last handshake is not older than some constant, instead of how it is now?

So nodes that have an outdated endpoint will eventually forget about it and get the updated endpoint from another nodes' annotation?

Reply:
That would be a great feature to add eventually

Reply:
The bare minimum is when `latest-handshake` is 0 wireguard didn't see any connection yet, so better not advertise it (anyway in that cas it's still equal to n.Endpoint). Then yes we could detect outdated endpoint.

Also in topology instead of a static ordering we could sort by `latest-handshake`, the fact that we give priority to the current detected endpoint already ensure stability, but having more _noise_ would allow to try the different endpoints one node could have (in case of really complicated network)

The cost however in another call to `wg` (ie `wg show kilo0 latest-handshakes`) as the `wg showconf` call will not return the `latest-handshake`